engine/src/main/java/org/terasology/rendering/nui/editor/NUIEditorContextMenuBuilder.java

```diff
@@ -0,0 +1,350 @@
+                addContextMenuListeners.forEach(UpdateListener::onAction);
+            }, node, true);
+
+            // Populate the menu with additional options.
+            if (node.hasChildWithKey("type")) {
+                populateWidgetContextMenu(node, addLevel);
+            } else if (node.getValue().getKey().equalsIgnoreCase("layoutInfo")) {
+                populateLayoutContextMenu(node, addLevel);
+            } else {
+                populateGenericContextMenu(node, addLevel);
+            }
+        }
+    }
+
+    private void populateWidgetContextMenu(JsonTree node, ContextMenuLevel addLevel) {
+        // Get the ClassMetadata of the widget with the given type.
+        String type = (String) node.getChildWithKey("type").getValue().getValue();
+        ClassMetadata<? extends UIWidget, ?> elementMetadata = nuiManager
+            .getWidgetMetadataLibrary()
+            .resolve(type, ModuleContext.getContext());
+
+        if (elementMetadata != null) {
+            for (FieldMetadata fieldMetadata : elementMetadata.getFields()) {
+                Field field = fieldMetadata.getField();
+                field.setAccessible(true);
+
+                // Add options for all @LayoutConfig-annotated fields.
+                if (field.isAnnotationPresent(LayoutConfig.class)) {
+                    String name = getNodeName(field);
+                    if (!node.hasChildWithKey(name)) {
+                        addLevel.addOption(name, n -> {
+                            try {
+                                createChild(name, node, field, elementMetadata);
+                            } catch (IllegalAccessException e) {
+                                return;
+                            }
+                            addContextMenuListeners.forEach(UpdateListener::onAction);
+                        }, node, true);
+                    }
+                }
+            }
+        } else {
+            logger.warn("Unknown widget type: {}", type);
+        }
+    }
+
+    private void populateLayoutContextMenu(JsonTree node, ContextMenuLevel addLevel) {
+        String type = "";
+        JsonTree currentNode = node;
+        ClassMetadata layoutMetadata = null;
+        Class layoutHintType = null;
+
+        // Iterate down the tree and attempt to find a UILayout node.
+        while (layoutHintType == null) {
+            if (currentNode.hasSiblingWithKey("type")) {
+                type = (String) currentNode.getSiblingWithKey("type").getValue().getValue();
+                layoutMetadata = nuiManager.getWidgetMetadataLibrary().resolve(type, ModuleContext.getContext());
+
+                // If one is found, attempt to get the layout hint type.
+                try {
+                    layoutHintType = (Class) ReflectionUtil
+                        .getTypeParameter(layoutMetadata.getType().getGenericSuperclass(), 0);
+                } catch (NullPointerException e) {
```

Comment:
Catching a NullPointerException is really bad design. There should be a null check at the proper place.

Reasoning: 
1. it is then clearer what is null. 
2. Exceptions are very costly in java, 
3. when nullpointerexceptions are only used for errors then it is possible to set a breakpoint on NullPointerExceptions and have the program only stop there and not at hunderts of other places